cmail-admin/doc/cmail-admin-smtpd.8
.TH CMAIL-ADMIN-SMTPD 8 "August 2015" "v0.1" "cmail Administration Tools"
.SH NAME
cmail-admin-smtpd \- Configure origination routing for cmail-smtpd
.SH SYNOPSIS
.BI "cmail-admin-smtpd [ --dbpath " "master-database-file" " ] [ --verbosity " N
.BI " | -v " N " ] [ --help | -h ] " "mode" " "  "mode-options"
.SH DESCRIPTION
.B cmail-admin-smtpd
allows the administrator of a cmail system to configure how originated mail
(mail that is being sent by a local, authenticated user) is handled
by the SMTP daemon.

.SH OPTIONS
.TP 
.BI "--dbpath " master-database-file ", -d " master-database-file
Select an alternate master database file instead of the default
.IR /etc/cmail/databases/master.db3 .
This overrides the database path provided by the
.B CMAIL_MASTER_DB
environment variable.
.TP
.BI "--verbose " N ", -v " N
Increase verbosity on a scale of
.B 0
to 
.B 4
.TP
.B "--help, -h"
Print helpful hints about operating the program
.SH FILES
.I /etc/cmail/databases/master.db3
.RS
Default path of the master database when installed system-wide. May be overridden either by specifying the
.B CMAIL_MASTER_DB
environment variable or by supplying the
.BR "--dbpath " parameter.
.SH MODES

.TP
.BI "list " [expression]
List all users currently able to authenticate with the SMTP daemon. 
Optionally, filter the list for user names matching
.IR expression "."

.TP
.BI "enable " "username [router [router-argument] ]"
.RI "Add " username " to the ACL, thus allowing SMTP authentication for this user. Users must exist and" 
have valid login credentials to successfully authenticate with the smtpd. User management tasks may for example 
be performed with
.BR cmail-admin-user (8)
.RS
.B Example:
$ cmail-admin-smtpd enable cbdev any
.RE

.TP
.BI "update " "user router [router-argument]"
Update an existing entry in the ACL with new routing data.
.RS
.B Example:
$ cmail-admin-smtpd update cbdev reject "No more mail, please"
.RE

.TP
.BI "disable " user
Remove a user from the ACL. Note that this does not prevent the user from receiving mail,
only from authenticating with the SMTP daemon and originating it.
.RS
.B Example:
$ cmail-admin-smtpd disable cbdev
.RE

.SH ROUTERS
Outbound routers are filters limiting the paths which a user may use as reverse paths
.RB "in originated mails (that is, they limit the addresses a user may " send " from)."

.TP
.BI any
Accept any reverse path

.TP
.BI defined
.RB "Accept only reverse paths that have a " store " router pointing back to the authorized"
.RB "user. Note that while there needs to be a " store " router set to the user, it need not be"
the one actually taken when the path is routed inbound.

.TP
.BI "handoff " remote-spec
.RB "Have originated mail by this user handled by another SMTP server " (smarthost) "."
This effectively relays the entire envelope section verbatim to the remote server.
.RI "If the " remote-spec " is not set, this router behaves the same way as"
.RB "the " drop " router. If"
.RI "the " remote-spec " is invalid,  the outbound smtp client will generate a bounce notification"
message after a configured amount of retries.

.TP
.BI "reject " [reason]
Reject any reverse path (SMTP 551), optionally presenting
.IR reason " as failure reason."

.TP
.BI drop
Accept any reverse path, but quietly drop all originated mail for this user.

.SH BUGS
For a discussion of pitfalls and possible sources of misconfigurations refer to the CAVEATS.txt file in the source distribution.
.SH AUTHORS
 Jan "mpease" Duepmeier <me at mpease dot de>  
 Fabian "cbdev" Stumpf <fjs at fabianstumpf dot de>  
.SH "SEE ALSO"
Online documentation and homepage at http://cmail.rocks/
.PP
.BR cmail-admin-address (8),
.BR cmail-admin-user (8),
.BR cmail-admin-popd (8),
.BR cmail-smtpd (8),
.BR cmail-popd (8),
.BR cmail-dispatchd (8),
.BR cmail-master-database (5)



debian/changelog
cmail-mta (1.0) unstable; urgency=low

  * Initial Release.

 -- root <root@unknown>  Sat, 18 Jul 2015 14:24:27 +0000


cmail-admin/smtpd/cmail-admin-smtpd.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sqlite3.h>
#include <time.h>

#include "../cmail-admin.h"
#include "smtpd.c"
#define PROGRAM_NAME "cmail-admin-smtpd"

int usage(char* fn) {
	printf("cmail-admin-smtpd - Manage cmail-smtpd ACL\n");
	printf("This program is part of the cmail administration toolkit\n");
	printf("Usage: %s <options> <commands> <arguments>\n\n", fn);

	printf("Basic options:\n");
	printf("\t--verbosity, -v\t\tSet verbosity level (0 - 4)\n");
	printf("\t--dbpath, -d <dbpath>\tSet master database path (Default: %s)\n", DEFAULT_DBPATH);
	printf("\t--help, -h\t\tDisplay this help message\n\n");

	printf("Commands:\n");
	printf("\tlist [<user>]\t\t\t\t\tList currently active entries, optionally filtered by <user>\n");
	printf("\tenable <user> [<router> [<router-argument>]]\tAdd <user> to smtpd ACL\n");
	printf("\tupdate <user> <router> [<router-argument>]\tUpdate the origination router for <user>\n");
	printf("\tdisable <user>\t\t\t\t\tDelete <user> from smtpd ACL\n");

	return 10;
}

int mode_add(LOGGER log, sqlite3* db, int argc, char** argv) {
	if (argc < 2) {
		logprintf(log, LOG_ERROR, "Missing arguments\n");
		return -1;
	}

	return sqlite_add_msa(log, db, argv[1], (argc > 2) ? argv[2]:NULL, (argc > 3) ? argv[3]:NULL);
}

int mode_delete(LOGGER log, sqlite3* db, int argc, char** argv) {
	if (argc != 2) {
		logprintf(log, LOG_ERROR, "No user name supplied\n");
		return -1;
	}

	return sqlite_delete_msa(log, db, argv[1]);
}

int mode_list(LOGGER log, sqlite3* db, int argc, char** argv) {
	char* filter = "%";

	if (argc >= 2) {
		filter = argv[1];
	}

	return sqlite_get_msa(log, db, filter);
}

int mode_update(LOGGER log, sqlite3* db, int argc, char** argv) {
	if (argc < 3 ) {
		logprintf(log, LOG_ERROR, "Missing arguments\n");
		return -1;
	}

	return sqlite_update_msa(log, db, argv[1], argv[2], (argc > 3) ? argv[3]:NULL);
}

// common stuff for parsing (needs usage method above)
#include "../lib/common.c"

int main(int argc, char* argv[]) {
	struct config config = {
		.verbosity = 0,
		.dbpath = getenv("CMAIL_MASTER_DB")
	};

	if (!config.dbpath) {
		config.dbpath = DEFAULT_DBPATH;
	}

	// argument parsing
	add_args();

	char* cmds[argc];
	int cmdsc = eargs_parse(argc, argv, cmds, &config);

	// check for errors in parsing
	if (cmdsc < 0) {
		return 1;
	}

	LOGGER log = {
		.stream = stderr,
		.verbosity = config.verbosity
	};

	sqlite3* db = NULL;

	// check for commands
	if (!cmdsc) {
		logprintf(log, LOG_ERROR, "No command specified\n\n");
		exit(usage(argv[0]));
	}

	logprintf(log, LOG_INFO, "Opening database at %s\n", config.dbpath);
	db = database_open(log, config.dbpath, SQLITE_OPEN_READWRITE);

	// check for database opening errors
	if (!db) {
		exit(usage(argv[0]));
	}

	//check database version
	if (database_schema_version(log, db) != CMAIL_CURRENT_SCHEMA_VERSION) {
		logprintf(log, LOG_ERROR, "The specified database (%s) is at an unsupported schema version.");
		sqlite3_close(db);
		return 11;
	}

	int status = 20;

	if (!strcmp(cmds[0], "enable")) {
		status = mode_add(log, db, cmdsc, cmds);
	}  else if (!strcmp(cmds[0], "disable")) {
		status = mode_delete(log, db, cmdsc, cmds);
	}  else if (!strcmp(cmds[0], "update")) {
		status = mode_update(log, db, cmdsc, cmds);
	}  else if (!strcmp(cmds[0], "list")) {
		status = mode_list(log, db, cmdsc, cmds);
	} else {

		logprintf(log, LOG_WARNING, "Unkown command %s\n\n", cmds[0]);
		usage(argv[0]);
	}

	sqlite3_close(db);

	return status;
}
